x/pairing/keeper/msg_server_relay_payment.go

```diff
@@ -392,3 +391,83 @@ func appendRelayPaymentDetailsToEvent(from map[string]string, uniqueIdentifier u
 	}
 	return to
 }
+
+// distributeRewards is the main function for reward distribution for providers and delegators
+func (k Keeper) distributeRewards(ctx sdk.Context, providerAddr sdk.AccAddress, chainID string, epoch uint64, totalReward math.Int) error {
+	providerReward, err := k.calculateProviderDelegatorsRewards(ctx, providerAddr, chainID, totalReward)
+	if err != nil {
+		return utils.LavaFormatError("cannot calculate provider reward and update delegator reward map", err)
+	}
+
+	providerRewardCoins := sdk.Coins{sdk.NewCoin(epochstoragetypes.TokenDenom, providerReward)}
+	err = k.bankKeeper.SendCoinsFromModuleToAccount(ctx, types.ModuleName, providerAddr, providerRewardCoins)
+	if err != nil {
+		// panic:ok: reward transfer should never fail
+		utils.LavaFormatPanic("critical: failed to send reward to provider", err,
+			utils.Attribute{Key: "provider", Value: providerAddr},
+			utils.Attribute{Key: "reward", Value: providerRewardCoins},
+		)
+	}
+
+	delegations, err := k.dualStakingKeeper.GetProviderDelegators(ctx, providerAddr.String(), epoch)
+	if err != nil {
+		return utils.LavaFormatError("cannot get provider's delegators", err)
+	}
+
+	err = k.updateDelegatorsReward(ctx, providerAddr, delegations, chainID, totalReward)
+	if err != nil {
+		return utils.LavaFormatError("cannot update delegators reward map", err)
+	}
+
+	return nil
+}
+
+// calculateProviderDelegatorsRewards returns the provider reward amount and updates the delegator reward map
+func (k Keeper) calculateProviderDelegatorsRewards(ctx sdk.Context, provider sdk.AccAddress, chainID string, totalReward math.Int) (math.Int, error) {
+	stakeEntry, found, _ := k.epochStorageKeeper.GetStakeEntryByAddressCurrent(ctx, chainID, provider)
```

Comment:
get the stake for the appropriate epoch

Reply:
done. see https://github.com/lavanet/lava/commit/4734092858e3f63ac104d340e6ed56a308a25f3c